dot/parachain/availability-distribution/fetch_chunk_task.go

```diff
@@ -0,0 +1,203 @@
+package availabilitydistribution
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	availabilitystore "github.com/ChainSafe/gossamer/dot/parachain/availability-store"
+	networkbridgemessages "github.com/ChainSafe/gossamer/dot/parachain/network-bridge/messages"
+	parachaintypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+)
+
+type taskTerminationReason uint
+
+const (
+	taskSucceeded taskTerminationReason = iota
+	taskFailed
+	taskCancelled
+)
+
+type taskTerminationHandler func(
+	candidateHash parachaintypes.CandidateHash,
+	reason taskTerminationReason,
+	badValidators []parachaintypes.AuthorityDiscoveryID,
+)
+
+type fetchChunkTask struct {
+	chunkIndex uint32
+	ourIndex   parachaintypes.ValidatorIndex
+	core       *parachaintypes.OccupiedCore
+	group      []parachaintypes.AuthorityDiscoveryID
+	overseerCh chan<- any
+
+	// Set of relay chain block hashes for which the candidate associated with the given core is pending availability.
+	//
+	// In other words, for which relay chain parents this candidate is considered live.
+	// This is updated on every `ActiveLeavesUpdate` and enables us to know when we can safely
+	// stop keeping track of that candidate/chunk.
+	liveIn map[common.Hash]struct{}
+	mu     sync.Mutex
+
+	// List of validators that did not have the requested chunk or sent invalid data.
+	badValidators []parachaintypes.AuthorityDiscoveryID
+
+	onTermination taskTerminationHandler
+
+	ctx    context.Context
+	cancel context.CancelFunc
```

Comment:
why do we need these?

Reply:
Tasks need to be cancelled when the node shuts down or when all leaves that the task is "live in" have been deactivated. The latter happens in the code for #4492 that I'm currently working on.

Reply:
can we not just use basic channels? Not sure why we need to use a `context.Context`.

Reply:
I've replaced `context.Context` with a channel in [this commit](https://github.com/ChainSafe/gossamer/pull/4618/commits/eeb8b3dfb424ed8f2cca6f5c2d0993e580c8fa59).